HCP Terraform Docs (#2835)
hashicorp/web-unified-docs: content/terraform-docs-common/docs/cloud-docs/workspaces/import.mdx

---
page_title: Search and import resources
description: Search your existing infrastructure for unmanaged resources and import them into your workspace. Learn how to discover and import infrastructure at scale.
---

# Search and import resources

> HCP Terraform lets you search your existing infrastructure for resources and import them into your workspace. This feature helps you discover unmanaged resources across cloud providers, making it easier to bring existing infrastructure under Terraform management.

Reviewer: ```suggestion
HCP Terraform and Terraform Enterprise let you search your existing infrastructure for resources and import them into your workspace. This feature helps you discover unmanaged resources across cloud providers, making it easier to bring existing infrastructure under Terraform management.
```
Just because this feature is in both editions. It's probably enough to mention TFE here and leave references to HCPTF throughout the rest of the topic. 

For extra points, we could also use the exclusion tags to customize when product name is rendered, but we probably don't need that many points right now.

## Requirements

To run queries and import resources, you need:

- An HCP Terraform workspace with Terraform 1.14 or newer configured
- Cloud provider credentials for the providers you wish to query configured for your workspace
- Read and write permissions for the workspace

HCP Terraform identifies resources managed by other workspaces when the workspace uses Terraform v1.12 and newer.  

### HCP Terraform agents

If you use HCP Terraform agents for your runs, you must enable the `query` operation when starting the agent pool. Refer to [Install and run agents](/terraform/cloud-docs/agents/agents) for more information.

## Define queries

To query resources with HCP Terraform, you can either use the build query UI, or write queries directly into your Terraform configuration.

### Build a query in the UI

To query resources for your workspace using the buiild query UI:

1. Navigate to your workspace in HCP Terraform.
1. Select **Search & Import** from the left navigation.
1. Click the **+ New query** dropdown in the upper right.
1. Select the **Build query via UI** button.

> 1. On the **New query** page, select the providers you wish to import resources for. Remember that your workspace must have credentials configured for the providers you select.

Reviewer 2: Should we link to provider configuration docs? For Ui driven queries to run, this step is very important.

1. Click the **Select provider** button to continue.
1. To begin building your query, select the resource types you want to query for. You can include multiple resource types from different providers in a single query.
1. To filter your query to only match certain resources of type you selected, click the **+** icon to the right of your resource type.
1. You can either select pre-defined filters based upon the resource type, or filter any attribute that Terraform tracks for the resource type.
1. After you have defined all of the filters for your query, click the **Run query** button to run your query.

>   The **Start new query** modal appears with the **Generate starter configuration** option enabled by default. This option directs HCP Terraform to generate starter configuration for you.

Reviewer 2: For `Generate starter configuration` 

This is a performance tuning option, i think we can put a separate section  that helps explain what checking the checkbox does. Whichever option the user picks does alter the import process

If we leave it checked, queries will be slower because terraform will get the full resource config of each resource found. The user can copy the results and follow the current import process. 

If unchecked, it returns just the identity of the resource resulting in a faster query run. This does alter the import process, the user will need to save the import blocks to a file (e.g. import.tf), then run terraform plan -generate-config-out=import.tf to generate corresponding resource blocks

Author: I've added this as a note below the list for now. Open to other options, though.

Author: ^ Actually I moved it into the list, I think that covers it?

Reviewer 2: Can we add it to its own dedicated section and link it? I think we plan on linking the UI to the doc page when that modal shows up inform the user as well

1. Click **Confirm**. Terraform searches for resources according your query.

### Define a query in Terraform configuration

To query resources for your workspace using Terraform configuration:

1. Add `list` blocks to your workspace's Terraform configuration to createthe search queries you want to run against your existing infrastructure. Refer to [Import resources in bulk](/terraform/language/import/bulk) for instructions on how to define queries. 
1. If your workspace uses the VCS-driven workflow, commit and push your changes to the branch configured in your workspace settings.
1. If your workspace uses the CLI-driven workflow, apply your changes from your terminal.
1. Navigate to your workspace in HCP Terraform.
1. Select **Search & Import** from the left navigation.
1. Click the **+ New query** dropdown in the upper right.
1. Select the **Fetch latest configuration** button.
1. Click **New Query**.

>     The **Start new query** modal appears with the **Generate starter configuration** option enabled by default. This option directs HCP Terraform to generate starter configuration for you.

Reviewer 2: this would be the same as this comment: https://github.com/hashicorp/web-unified-docs/pull/2835/changes#r3590646515

---
page_title: Search and import resources
description: Search your existing infrastructure for unmanaged resources and import them into your workspace. Learn how to discover and import infrastructure at scale.
---

# Search and import resources

HCP Terraform lets you search your existing infrastructure for resources and import them into your workspace. This feature helps you discover unmanaged resources across cloud providers, making it easier to bring existing infrastructure under Terraform management.

## Background

When you inherit infrastructure or manage environments with unknown resources, identifying and importing unmanaged resources can be a time-consuming manual process. The search and import feature provides a visual interface to build queries, select resources, and generate configuration so that HCP Terraform can discover resources across your cloud providers.

> ## Introduction

Reviewer: ```suggestion
### Workflow
```
This section seems to more like a subset of the background that describes the workflow.

HCP Terraform executes search queries against your cloud providers to discover resources that are not currently managed by Terraform. You can either define these queries in the HCP Terraform UI, or in your workspace's Terraform configuration.

When you query resources, the search interface displays results with management status indicators, helping you identify which resources to import. You can write or automatically generate Terraform configuration to import and manage the selected resources, and then apply your changes to bring them under management.

## Requirements

To run queries and import resources, you need:

- An HCP Terraform workspace with Terraform 1.14 or newer configured
- Cloud provider credentials for the providers you wish to query configured for your workspace

> - Read and write permissions for the workspace

Reviewer: Is this the same requires as line 34-38?

### Permissions

Your HCP Terraform user must have the following permissions to run queries and import resources:

- **Manage workspace** permission for the workspace

## Define queries

To query resources with HCP Terraform, you can either use the build query UI, or write queries directly into your Terraform configuration.

### Build a query in the UI

To query resources for your workspace using the buiild query UI:

1. Navigate to your workspace in HCP Terraform.
1. Select **Search & Import** from the left navigation.
1. Click the **+ New query** dropdown in the upper right.
1. Select the **Build query via UI** button.
1. On the **New query** page, select the providers you wish to import resources for. Remember that your workspace must have credentials configured for the providers you select.
1. Click the **Select provider** button to continue.
1. To begin building your query, select the resource types you want to query for. You can include multiple resource types from different providers in a single query.
1. To filter your query to only match certain resources of type you selected, click the **+** icon to the right of your resource type.

> 1. You can select pre-defined filters based upon the resource type, or filter any attribute that Terraform tracks for the resource type.

Reviewer: ```suggestion
1. You can either select pre-defined filters based upon the resource type, or filter any attribute that Terraform tracks for the resource type.
```

1. After you have defined all of the filters for your query, click the **Run query** button to run your query.

>   A **Start new query** modal will pop up. To have HCP Terraform generate starter configuration for you, leave the **Generate starter configuration** box checked.

Reviewer: ```suggestion
  The **Start new query** modal appears with the **Generate starter configuration** option enabled by default. This option directs HCP Terraform to generate starter configuration for you.
```

> 1. Click **Confirm** to have HCP Terraform query your resources.

Reviewer: ```suggestion
1. Click **Confirm**. Terraform searches for resources according your query.
```

> ### Configure a query

Reviewer: ```suggestion
### Define a query in Terraform configuration
```
Trying to use a heading that more clearly differentiates from using the UI. Ignore if this doesn't seem like an improvement.

To query resources for your workspace using Terraform configuration:

1. Add `list` blocks to your workspace's Terraform configuration to createthe search queries you want to run against your existing infrastructure. Refer to [Import resources in bulk](/terraform/language/import/bulk) for instructions on how to define queries. 
1. If your workspace uses the VCS-driven workflow, commit and push your changes to the branch configured in your workspace settings.
1. If your workspace uses the CLI-driven workflow, apply your changes from your terminal.
1. Navigate to your workspace in HCP Terraform.
1. Select **Search & Import** from the left navigation.
1. Click the **+ New query** dropdown in the upper right.
1. Select the **Fetch latest configuration** button.
1. Click **New Query**.
   A **Start new query** modal will pop up. To have HCP Terraform generate starter configuration for you, leave the **Generate starter configuration** box checked.

> 1. Click **Confirm** to have HCP Terraform query your resources.

Reviewer: ```suggestion
    The **Start new query** modal appears with the **Generate starter configuration** option enabled by default. This option directs HCP Terraform to generate starter configuration for you.
1. Click **Confirm**. Terraform searches for resources according your query.
```

## Review results

HCP Terraform displays the query results, including the IaC status, if known.

> The **IaC Status** colum will report whether the resource matches one already being managed by HCP Terraform:

Reviewer: ```suggestion
The **IaC Status** column indicates when a discovered resource matches a resource already managed by HCP Terraform:
```

If your workspace uses the CLI-driven workflow, plan and apply your changes from your terminal.

You may need to modify the generated configuration to succesffully import your resources. Resolve any errors in the Terraform plan, and then apply your changes.

## Next steps

After importing the resources to state, you can delete the generated import blocks or keep them as a historical record.  

### Reference

Refer to the following documentation to learn more about how to import resources with Terraform:

- [Import resources in bulk](/terraform/language/import/bulk)
- [Import a single resource](/terraform/language/import/single-resource)
- [Workspace permissions](/terraform/cloud-docs/users-teams-organizations/permissions)

### Constraints and limitations

- You must enable Terraform 1.14 or newer for your workspace to access the search and import feature.
- Search queries require appropriate cloud provider credentials configured in your workspace.
- If you use HCP Terraform agents, you must enable the `query` operation when starting the agent pool.

> - Resources managed by other workspaces are identified only when those workspaces use Terraform v1.12 or newer.

Reviewer: We should move this to the Requirements section per the [content type guidance](https://hashicorp.atlassian.net/wiki/spaces/TW/pages/2786951262/Design+for+usage+pages)

Author: Good catch, those are duplicates - the "Resources managed by other workspaces ... Terraform v1.12" seems like an actual limitation. We do mention it in the query section, but should we keep it here as well?

Reviewer: If we're splitting hairs, yeah, that line isn't really a "Requirements", but the spec calls for putting all of this kind of information near the beginning of the topic is so that readers 1) have it all in one place and 2) aren't surprised when they get to the bottom.

We haven't been super strict on this guidance, though, especially when the list of limitations bloat the Requirements section. Your call.

1. Select the **Build query via UI** button.
1. On the **New query** page, select the providers you wish to import resources for. Remember that your workspace must have credentials configured for the providers you select.
1. Click the **Select provider** button to continue.
1. To begin building your query, select the resource types you want to query for. You can include multiple resource types from different providers in a single query.
1. To filter your query to only match certain resources of type you selected, click the **+** icon to the right of your resource type.
1. You can either select pre-defined filters based upon the resource type, or filter any attribute that Terraform tracks for the resource type.
1. After you have defined all of the filters for your query, click the **Run query** button to run your query.
  The **Start new query** modal appears with the **Generate starter configuration** option enabled by default. This option directs HCP Terraform to generate starter configuration for you by loading additional data about the resources it queries.
1. Click **Confirm**. Terraform searches for resources according your query.

<Note>

> When you enable the **Generate starter configuration** option, HCP Terraform retrieves additonal data about your resources in order to generate the configuration. Queries with this option enabled will take longer to complete.

Reviewer: ```suggestion
When you enable the **Generate starter configuration** option, HCP Terraform retrieves additonal data about your resources in order to generate the configuration. Queries with this option enabled take longer to complete.
```
